File: zig/src/engine.zig
//! The construction over the Threefish cipher: the authenticated chunked password
//! container, raw CTR, and inspect. Streams over Reader/Writer callbacks in constant
//! memory (files larger than RAM are fine), matching the Rust reference. Slice
//! convenience wrappers are provided. The on-disk output is byte-for-byte identical
//! to the other ports, so `.mahi` files are cross-compatible.
//!
//! Educational and unaudited.

const std = @import("std");
const mem = std.mem;
const tf = @import("threefish.zig");
const fmt = @import("format.zig");
const kdf = @import("kdf.zig");
const mac_mod = @import("mac.zig");

pub const Variant = tf.Variant;
pub const Kdf = fmt.Kdf;
pub const Mac = fmt.Mac;
pub const Header = fmt.Header;

const FRAME_DOMAIN = "DRDOchnk";
const HEADER_MAX = 8192;

pub const Reader = fmt.Reader;

/// A byte sink: returns true on success.
pub const Writer = struct {
    ctx: *anyopaque,
    writeFn: *const fn (ctx: *anyopaque, bytes: []const u8) bool,

    pub fn write(self: Writer, bytes: []const u8) Error!void {
        if (!self.writeFn(self.ctx, bytes)) return Error.WriteFailed;
    }
};

pub const Error = error{
    LabelTooLong,
    InvalidChunkSize,
    BadContainer,
    AuthFailed,
    Truncated,
    WriteFailed,
    OutOfMemory,
    Rng,
    KdfFailed,
    HostileCost,
} || fmt.Error || kdf.Error;

pub const Options = struct {
    variant: Variant = .t256,
    kdf: Kdf = Kdf.mkArgon2id(64 * 1024, 3, 1),
    mac: Mac = .skein,
    tweak: [16]u8 = [_]u8{0} ** 16,
    chunk_size: u32 = fmt.DEFAULT_CHUNK_BYTES,
    label: []const u8 = "",
};

pub const ContainerInfo = struct {
    version: u8,
    variant: Variant,
    mac: Mac,
    kdf: Kdf,
    chunk_size: u32,
    salt_len: usize,
    tweak: [16]u8,
    label: [fmt.MAX_LABEL_LEN]u8,
    label_len: usize,

    pub fn labelSlice(self: *const ContainerInfo) []const u8 {
        return self.label[0..self.label_len];
    }
};

fn writeBE32(buf: []u8, v: u32) void {
    mem.writeInt(u32, buf[0..4], v, .big);
}

// Build the AAD for a frame into `dst`, returning the length used.
fn buildAad(dst: []u8, header_bytes: []const u8, index: u64, is_last: bool, ct: []const u8) usize {
    var n: usize = 0;
    @memcpy(dst[0..8], FRAME_DOMAIN);
    n += 8;
    if (index == 0) {
        @memcpy(dst[n .. n + header_bytes.len], header_bytes);
        n += header_bytes.len;
    }
    mem.writeInt(u64, dst[n..][0..8], index, .big);
    n += 8;
    dst[n] = if (is_last) 1 else 0;
    n += 1;
    writeBE32(dst[n..], @intCast(ct.len));
    n += 4;
    @memcpy(dst[n .. n + ct.len], ct);
    n += ct.len;
    return n;
}

fn writeFrame(w: Writer, is_last: bool, ct: []const u8, tag: *const [32]u8) Error!void {
    var head: [5]u8 = undefined;
    head[0] = if (is_last) 1 else 0;
    writeBE32(head[1..], @intCast(ct.len));
    try w.write(&head);
    if (ct.len > 0) try w.write(ct);
    try w.write(tag);
}

pub fn encryptStream(
    allocator: mem.Allocator,
    io: std.Io,
    password: []const u8,
    opts: Options,
    r: Reader,
    w: Writer,
) Error!void {
    if (opts.label.len > fmt.MAX_LABEL_LEN) return Error.LabelTooLong;
    const v = opts.variant;
    const bl = v.keyLen();
    if (opts.chunk_size == 0 or opts.chunk_size > fmt.MAX_CHUNK_BYTES or opts.chunk_size % bl != 0)
        return Error.InvalidChunkSize;

    var salt: [16]u8 = undefined;
    var iv: [128]u8 = undefined;
    io.randomSecure(&salt) catch return Error.Rng;
    io.randomSecure(iv[0..bl]) catch return Error.Rng;

    var keymat: [160]u8 = undefined;
    try kdf.derive(allocator, io, opts.kdf, password, &salt, keymat[0 .. bl + 32]);
    const enc_key = keymat[0..bl];
    const mac_key = keymat[bl .. bl + 32][0..32];

    var h: Header = undefined;
    h.version = fmt.VERSION;
    h.variant = v;
    h.kdf = opts.kdf;
    h.mac = opts.mac;
    h.chunk_size = opts.chunk_size;
    @memcpy(h.salt[0..16], &salt);
    h.salt_len = 16;
    h.tweak = opts.tweak;
    @memcpy(h.iv[0..bl], iv[0..bl]);
    @memcpy(h.label[0..opts.label.len], opts.label);
    h.label_len = opts.label.len;

    var hb: [HEADER_MAX]u8 = undefined;
    const hb_len = fmt.marshal(&h, &hb);

    var cipher = tf.Threefish.init(v, enc_key, &opts.tweak);
    var ctr = cipher.newCtr(iv[0..bl]);

    try w.write(hb[0..hb_len]);

    const cs = opts.chunk_size;
    const cur = try allocator.alloc(u8, cs);
    defer allocator.free(cur);
    const nxt = try allocator.alloc(u8, cs);
    defer allocator.free(nxt);
    const aad = try allocator.alloc(u8, 8 + hb_len + 13 + cs);
    defer allocator.free(aad);

    var cur_buf = cur;
    var nxt_buf = nxt;
    var cur_len = readSome(r, cur_buf);
    var index: u64 = 0;
    while (true) {
        const nxt_len = readSome(r, nxt_buf);
        const is_last = nxt_len == 0;
        ctr.apply(cur_buf[0..cur_len]);
        var tag: [32]u8 = undefined;
        const aad_len = buildAad(aad, hb[0..hb_len], index, is_last, cur_buf[0..cur_len]);
        mac_mod.tag(opts.mac, mac_key, aad[0..aad_len], &tag);
        try writeFrame(w, is_last, cur_buf[0..cur_len], &tag);
        if (is_last) break;
        index += 1;
        const tmp = cur_buf;
        cur_buf = nxt_buf;
        nxt_buf = tmp;
        cur_len = nxt_len;
    }
}

// Read up to buf.len bytes; a short count means end of input.
fn readSome(r: Reader, buf: []u8) usize {
    var got: usize = 0;
    while (got < buf.len) {
        const n = r.readFn(r.ctx, buf[got..]);
        if (n == 0) break;
        got += n;
    }
    return got;
}

pub fn decryptStream(
    allocator: mem.Allocator,
    io: std.Io,
    password: []const u8,
    expected_label: ?[]const u8,
    r: Reader,
    w: Writer,
) Error!void {
    var h: Header = undefined;
    try fmt.read(r, &h);
    if (expected_label) |el| {
        if (!mem.eql(u8, el, h.labelSlice())) return Error.BadContainer;
    }
    const bl = h.variant.keyLen();
    if (h.chunk_size == 0 or h.chunk_size > fmt.MAX_CHUNK_BYTES or h.chunk_size % bl != 0)
        return Error.InvalidChunkSize;
    try kdf.validate(h.kdf);

    var keymat: [160]u8 = undefined;
    try kdf.derive(allocator, io, h.kdf, password, h.saltSlice(), keymat[0 .. bl + 32]);
    const enc_key = keymat[0..bl];
    const mac_key = keymat[bl .. bl + 32][0..32];

    var hb: [HEADER_MAX]u8 = undefined;
    const hb_len = fmt.marshal(&h, &hb);

    var cipher = tf.Threefish.init(h.variant, enc_key, &h.tweak);
    var ctr = cipher.newCtr(h.ivSlice());

    const cs = h.chunk_size;
    const ct = try allocator.alloc(u8, cs);
    defer allocator.free(ct);
    const aad = try allocator.alloc(u8, 8 + hb_len + 13 + cs);
    defer allocator.free(aad);

    var index: u64 = 0;
    while (true) {
        var head: [5]u8 = undefined;
        const got = readSome(r, &head);
        if (got == 0) return Error.Truncated;
        if (got < 5) return Error.Truncated;
        if (head[0] > 1) return Error.BadContainer;
        const is_last = head[0] == 1;
        const ct_len = mem.readInt(u32, head[1..5], .big);
        if (ct_len > cs) return Error.BadContainer;
        if (!r.readFull(ct[0..ct_len])) return Error.Truncated;
        var tag: [32]u8 = undefined;
        if (!r.readFull(&tag)) return Error.Truncated;
        const aad_len = buildAad(aad, hb[0..hb_len], index, is_last, ct[0..ct_len]);
        if (!mac_mod.verify(h.mac, mac_key, aad[0..aad_len], &tag)) return Error.AuthFailed;
        ctr.apply(ct[0..ct_len]);
        if (ct_len > 0) try w.write(ct[0..ct_len]);
        if (is_last) break;
        if (ct_len != cs) return Error.BadContainer;
        index += 1;
    }
}

pub fn rawCtrStream(variant: Variant, key: []const u8, tweak: []const u8, iv: []const u8, allocator: mem.Allocator, r: Reader, w: Writer) Error!void {
    const bl = variant.keyLen();
    var cipher = tf.Threefish.init(variant, key, tweak);
    var ctr = cipher.newCtr(iv);
    const bufsz = (65536 / bl) * bl;
    const buf = try allocator.alloc(u8, bufsz);
    defer allocator.free(buf);
    while (true) {
        const n = readSome(r, buf);
        if (n == 0) break;
        ctr.apply(buf[0..n]);
        try w.write(buf[0..n]);
        if (n < bufsz) break;
    }
}

pub fn inspectStream(r: Reader, info: *ContainerInfo) Error!void {
    var h: Header = undefined;
    try fmt.read(r, &h);
    info.version = h.version;
    info.variant = h.variant;
    info.mac = h.mac;
    info.kdf = h.kdf;
    info.chunk_size = h.chunk_size;
    info.salt_len = h.salt_len;
    info.tweak = h.tweak;
    @memcpy(info.label[0..h.label_len], h.labelSlice());
    info.label_len = h.label_len;
}

// ---- slice convenience wrappers (over a fixed buffer / arraylist) ----

const SliceReader = struct {
    data: []const u8,
    pos: usize = 0,

    fn read(ctx: *anyopaque, buf: []u8) usize {
        const self: *SliceReader = @ptrCast(@alignCast(ctx));
        const n = @min(buf.len, self.data.len - self.pos);
        @memcpy(buf[0..n], self.data[self.pos .. self.pos + n]);
        self.pos += n;
        return n;
    }

    fn reader(self: *SliceReader) Reader {
        return .{ .ctx = self, .readFn = read };
    }
};

const ListWriter = struct {
    list: *std.ArrayList(u8),
    allocator: mem.Allocator,
    ok: bool = true,

    fn write(ctx: *anyopaque, bytes: []const u8) bool {
        const self: *ListWriter = @ptrCast(@alignCast(ctx));
        self.list.appendSlice(self.allocator, bytes) catch {
            self.ok = false;
            return false;
        };
        return true;
    }

    fn writer(self: *ListWriter) Writer {
        return .{ .ctx = self, .writeFn = write };
    }
};

/// Encrypt plaintext into a freshly-allocated container (caller frees).
pub fn encrypt(allocator: mem.Allocator, io: std.Io, password: []const u8, opts: Options, plaintext: []const u8) Error![]u8 {
    var sr = SliceReader{ .data = plaintext };
    var list: std.ArrayList(u8) = .empty;
    errdefer list.deinit(allocator);
    var lw = ListWriter{ .list = &list, .allocator = allocator };
    try encryptStream(allocator, io, password, opts, sr.reader(), lw.writer());
    return list.toOwnedSlice(allocator);
}

/// Decrypt a container into a freshly-allocated buffer (caller frees).
pub fn decrypt(allocator: mem.Allocator, io: std.Io, password: []const u8, expected_label: ?[]const u8, data: []const u8) Error![]u8 {
    var sr = SliceReader{ .data = data };
    var list: std.ArrayList(u8) = .empty;
    errdefer list.deinit(allocator);
    var lw = ListWriter{ .list = &list, .allocator = allocator };
    try decryptStream(allocator, io, password, expected_label, sr.reader(), lw.writer());
    return list.toOwnedSlice(allocator);
}

pub fn inspect(data: []const u8, info: *ContainerInfo) Error!void {
    var sr = SliceReader{ .data = data };
    try inspectStream(sr.reader(), info);
}

File: c/src/engine.c
#include "dorado/engine.h"

#include <stdlib.h>
#include <string.h>
#include <sys/random.h>

#include "dorado/threefish.h"
#include "format.h"
#include "internal.h"
#include "kdf.h"
#include "mac.h"

dorado_kdf_params dorado_kdf_argon2id(uint32_t m_cost_kib, uint32_t t_cost, uint32_t p_cost) {
    dorado_kdf_params k = {0};
    k.kind = DORADO_KDF_ARGON2ID;
    k.m_cost = m_cost_kib;
    k.t_cost = t_cost;
    k.p_cost = p_cost;
    return k;
}

dorado_kdf_params dorado_kdf_scrypt(uint8_t log_n, uint32_t r, uint32_t p) {
    dorado_kdf_params k = {0};
    k.kind = DORADO_KDF_SCRYPT;
    k.log_n = log_n;
    k.r = r;
    k.p = p;
    return k;
}

dorado_kdf_params dorado_kdf_pbkdf2(uint32_t rounds) {
    dorado_kdf_params k = {0};
    k.kind = DORADO_KDF_PBKDF2;
    k.rounds = rounds;
    k.prf = 1;
    return k;
}

dorado_options dorado_default_options(void) {
    dorado_options o = {0};
    o.variant = DORADO_T256;
    o.kdf = dorado_kdf_argon2id(64 * 1024, 3, 1);
    o.mac = DORADO_MAC_SKEIN;
    o.chunk_size = DORADO_DEFAULT_CHUNK_BYTES;
    return o;
}

static size_t build_aad(uint8_t *buf, const uint8_t *header_bytes, size_t header_len, uint64_t index, int is_last,
                        const uint8_t *ct, size_t ct_len) {
    size_t n = 0;
    memcpy(buf + n, "DRDOchnk", 8);
    n += 8;
    if (index == 0) {
        memcpy(buf + n, header_bytes, header_len);
        n += header_len;
    }
    dorado_store64_be(buf + n, index);
    n += 8;
    buf[n++] = is_last ? 1 : 0;
    dorado_store32_be(buf + n, (uint32_t)ct_len);
    n += 4;
    memcpy(buf + n, ct, ct_len);
    n += ct_len;
    return n;
}

static const char *write_frame(FILE *out, int is_last, const uint8_t *ct, size_t ct_len, const uint8_t tag[32]) {
    uint8_t head[5];
    head[0] = (uint8_t)(is_last ? 1 : 0);
    dorado_store32_be(head + 1, (uint32_t)ct_len);
    if (fwrite(head, 1, 5, out) != 5) {
        return "write error";
    }
    if (ct_len && fwrite(ct, 1, ct_len, out) != ct_len) {
        return "write error";
    }
    if (fwrite(tag, 1, 32, out) != 32) {
        return "write error";
    }
    return NULL;
}

const char *dorado_encrypt_password_stream(const uint8_t *password, size_t password_len, const dorado_options *opts,
                                           FILE *in, FILE *out) {
    if (opts->label_len > DORADO_MAX_LABEL_LEN) {
        return "label too long";
    }
    int v = opts->variant;
    int bl = dorado_variant_len(v);
    if (bl == 0) {
        return "unknown variant";
    }
    if (opts->chunk_size == 0 || opts->chunk_size > DORADO_MAX_CHUNK_BYTES || opts->chunk_size % (uint32_t)bl != 0) {
        return "chunk size must be a positive multiple of the block size";
    }
    uint8_t salt[16];
    uint8_t iv[128];
    if (getentropy(salt, 16) != 0 || getentropy(iv, (size_t)bl) != 0) {
        return "RNG failure";
    }
    int kl = bl;
    uint8_t keymat[160];
    const char *e = dorado_kdf_derive(&opts->kdf, password, password_len, salt, 16, keymat, (size_t)kl + 32);
    if (e) {
        return e;
    }

    dorado_header h;
    memset(&h, 0, sizeof h);
    h.version = DORADO_VERSION;
    h.variant = v;
    h.kdf = opts->kdf;
    h.mac = opts->mac;
    h.chunk_size = opts->chunk_size;
    memcpy(h.salt, salt, 16);
    h.salt_len = 16;
    memcpy(h.tweak, opts->tweak, 16);
    memcpy(h.iv, iv, (size_t)bl);
    if (opts->label_len) {
        memcpy(h.label, opts->label, opts->label_len);
    }
    h.label_len = opts->label_len;
    uint8_t hb[DORADO_HEADER_MAX_BYTES];
    size_t hb_len = dorado_format_marshal(&h, hb);

    dorado_threefish tf;
    dorado_threefish_init(&tf, v, keymat, opts->tweak);
    dorado_ctr ctr;
    dorado_ctr_init(&ctr, &tf, iv);

    if (fwrite(hb, 1, hb_len, out) != hb_len) {
        return "write error";
    }

    size_t cs = opts->chunk_size;
    uint8_t *cur = malloc(cs);
    uint8_t *nxt = malloc(cs);
    uint8_t *aad = malloc(8 + hb_len + 13 + cs);
    if (!cur || !nxt || !aad) {
        free(cur);
        free(nxt);
        free(aad);
        return "out of memory";
    }
    size_t cur_len = fread(cur, 1, cs, in);
    uint64_t index = 0;
    const char *err = NULL;
    for (;;) {
        size_t nxt_len = fread(nxt, 1, cs, in);
        int is_last = nxt_len == 0;
        dorado_ctr_apply(&ctr, cur, cur_len);
        uint8_t tag[32];
        size_t aad_len = build_aad(aad, hb, hb_len, index, is_last, cur, cur_len);
        dorado_mac_tag(opts->mac, keymat + kl, aad, aad_len, tag);
        err = write_frame(out, is_last, cur, cur_len, tag);
        if (err || is_last) {
            break;
        }
        index++;
        uint8_t *tmp = cur;
        cur = nxt;
        nxt = tmp;
        cur_len = nxt_len;
    }
    free(cur);
    free(nxt);
    free(aad);
    return err;
}

const char *dorado_decrypt_password_stream(const uint8_t *password, size_t password_len, const uint8_t *expected_label,
                                           size_t expected_label_len, FILE *in, FILE *out) {
    dorado_header h;
    const char *e = dorado_format_read(in, &h);
    if (e) {
        return e;
    }
    if (expected_label) {
        if (expected_label_len != h.label_len || memcmp(expected_label, h.label, expected_label_len) != 0) {
            return "container label does not match the expected label";
        }
    }
    int bl = dorado_variant_len(h.variant);
    if (h.chunk_size == 0 || h.chunk_size > DORADO_MAX_CHUNK_BYTES || h.chunk_size % (uint32_t)bl != 0) {
        return "invalid chunk size in header";
    }
    e = dorado_kdf_validate(&h.kdf);
    if (e) {
        return e;
    }
    int kl = bl;
    uint8_t keymat[160];
    e = dorado_kdf_derive(&h.kdf, password, password_len, h.salt, h.salt_len, keymat, (size_t)kl + 32);
    if (e) {
        return e;
    }
    uint8_t hb[DORADO_HEADER_MAX_BYTES];
    size_t hb_len = dorado_format_marshal(&h, hb);

    dorado_threefish tf;
    dorado_threefish_init(&tf, h.variant, keymat, h.tweak);
    dorado_ctr ctr;
    dorado_ctr_init(&ctr, &tf, h.iv);

    size_t cs = h.chunk_size;
    uint8_t *ct = malloc(cs);
    uint8_t *aad = malloc(8 + hb_len + 13 + cs);
    if (!ct || !aad) {
        free(ct);
        free(aad);
        return "out of memory";
    }
    uint64_t index = 0;
    const char *err = NULL;
    for (;;) {
        uint8_t head[5];
        size_t got = fread(head, 1, 5, in);
        if (got == 0) {
            err = "stream ended before the final chunk (truncated)";
            break;
        }
        if (got < 5) {
            err = "incomplete frame header (truncated)";
            break;
        }
        int flag = head[0];
        if (flag > 1) {
            err = "invalid frame flag";
            break;
        }
        int is_last = flag == 1;
        uint32_t ct_len = dorado_load32_be(head + 1);
        if (ct_len > cs) {
            err = "frame length exceeds the header chunk size";
            break;
        }
        if (!dorado_read_full(in, ct, ct_len)) {
            err = "truncated frame ciphertext";
            break;
        }
        uint8_t tag[32];
        if (!dorado_read_full(in, tag, 32)) {
            err = "truncated frame tag";
            break;
        }
        size_t aad_len = build_aad(aad, hb, hb_len, index, is_last, ct, ct_len);
        if (!dorado_mac_verify(h.mac, keymat + kl, aad, aad_len, tag)) {
            err = "authentication failed (wrong password, corruption, or tampering)";
            break;
        }
        dorado_ctr_apply(&ctr, ct, ct_len);
        if (ct_len && fwrite(ct, 1, ct_len, out) != ct_len) {
            err = "write error";
            break;
        }
        if (is_last) {
            break;
        }
        if (ct_len != cs) {
            err = "non-final chunk is not full size";
            break;
        }
        index++;
    }
    free(ct);
    free(aad);
    return err;
}

const char *dorado_raw_ctr_stream(int variant, const uint8_t *key, const uint8_t *tweak, const uint8_t *iv, FILE *in,
                                  FILE *out) {
    int bl = dorado_variant_len(variant);
    if (bl == 0) {
        return "unknown variant";
    }
    dorado_threefish tf;
    dorado_threefish_init(&tf, variant, key, tweak);
    dorado_ctr ctr;
    dorado_ctr_init(&ctr, &tf, iv);
    size_t bufsz = (size_t)(65536 / bl) * (size_t)bl;
    uint8_t *buf = malloc(bufsz);
    if (!buf) {
        return "out of memory";
    }
    const char *err = NULL;
    for (;;) {
        size_t n = fread(buf, 1, bufsz, in);
        if (n == 0) {
            break;
        }
        dorado_ctr_apply(&ctr, buf, n);
        if (fwrite(buf, 1, n, out) != n) {
            err = "write error";
            break;
        }
        if (n < bufsz) {
            break;
        }
    }
    free(buf);
    return err;
}

const char *dorado_inspect_stream(FILE *in, dorado_container_info *info) {
    dorado_header h;
    const char *e = dorado_format_read(in, &h);
    if (e) {
        return e;
    }
    info->version = h.version;
    info->variant = h.variant;
    info->mac = h.mac;
    info->kdf = h.kdf;
    info->chunk_size = h.chunk_size;
    info->salt_len = h.salt_len;
    memcpy(info->tweak, h.tweak, 16);
    memcpy(info->label, h.label, h.label_len);
    info->label_len = h.label_len;
    return NULL;
}

/* ---- in-memory wrappers (fmemopen / open_memstream) ---- */

const char *dorado_encrypt_password(const uint8_t *password, size_t password_len, const dorado_options *opts,
                                    const uint8_t *plaintext, size_t plaintext_len, uint8_t **out, size_t *out_len) {
    /* fmemopen with size 0 is unsupported on some platforms (macOS), so use an
     * empty tmpfile for empty input. */
    FILE *in = plaintext_len ? fmemopen((void *)plaintext, plaintext_len, "rb") : tmpfile();
    if (!in) {
        return "input stream failed";
    }
    char *buf = NULL;
    size_t sz = 0;
    FILE *of = open_memstream(&buf, &sz);
    if (!of) {
        fclose(in);
        return "open_memstream failed";
    }
    const char *err = dorado_encrypt_password_stream(password, password_len, opts, in, of);
    fclose(in);
    fclose(of);
    if (err) {
        free(buf);
        return err;
    }
    *out = (uint8_t *)buf;
    *out_len = sz;
    return NULL;
}

const char *dorado_decrypt_password(const uint8_t *password, size_t password_len, const uint8_t *expected_label,
                                    size_t expected_label_len, const uint8_t *data, size_t data_len, uint8_t **out,
                                    size_t *out_len) {
    FILE *in = fmemopen((void *)data, data_len, "rb");
    if (!in) {
        return "fmemopen failed";
    }
    char *buf = NULL;
    size_t sz = 0;
    FILE *of = open_memstream(&buf, &sz);
    if (!of) {
        fclose(in);
        return "open_memstream failed";
    }
    const char *err =
        dorado_decrypt_password_stream(password, password_len, expected_label, expected_label_len, in, of);
    fclose(in);
    fclose(of);
    if (err) {
        free(buf);
        return err;
    }
    *out = (uint8_t *)buf;
    *out_len = sz;
    return NULL;
}

const char *dorado_inspect(const uint8_t *data, size_t data_len, dorado_container_info *info) {
    FILE *in = fmemopen((void *)data, data_len, "rb");
    if (!in) {
        return "fmemopen failed";
    }
    const char *e = dorado_inspect_stream(in, info);
    fclose(in);
    return e;
}
